src/Storages/MergeTree/MergeTask.cpp

```diff
@@ -969,14 +997,16 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream()
     global_ctx->horizontal_stage_progress = std::make_unique<MergeStageProgress>(
         ctx->column_sizes ? ctx->column_sizes->keyColumnsWeight() : 1.0);
 
+    auto merging_column_names = global_ctx->merging_columns.getNames();
+
     for (const auto & part : global_ctx->future_part->parts)
     {
         Pipe pipe = createMergeTreeSequentialSource(
             MergeTreeSequentialSourceType::Merge,
             *global_ctx->data,
             global_ctx->storage_snapshot,
             part,
-            global_ctx->merging_column_names,
+            merging_column_names,
```

Comment:
```suggestion
    for (const auto & part : global_ctx->future_part->parts)
    {
        Pipe pipe = createMergeTreeSequentialSource(
            MergeTreeSequentialSourceType::Merge,
            *global_ctx->data,
            global_ctx->storage_snapshot,
            part,
            global_ctx->merging_columns.getNames(),
```

This way we won't copy `merging_column_names`.